components/tidb_query/src/rpn_expr/impl_cast.rs

```diff
@@ -124,17 +127,93 @@ pub fn get_cast_fn_rpn_node(
 
 fn produce_dec_with_specified_tp(
     ctx: &mut EvalContext,
-    dec: Decimal,
+    mut dec: Decimal,
     ft: &FieldType,
 ) -> Result<Decimal> {
-    // FIXME: The implementation is not exactly the same as TiDB's `ProduceDecWithSpecifiedTp`.
     let (flen, decimal) = (ft.flen(), ft.decimal());
-    if flen == tidb_query_datatype::UNSPECIFIED_LENGTH
-        || decimal == tidb_query_datatype::UNSPECIFIED_LENGTH
-    {
-        return Ok(dec);
+    if flen != UNSPECIFIED_LENGTH && decimal != UNSPECIFIED_LENGTH {
+        dec = dec.convert_to(ctx, flen as u8, decimal as u8)?;
+    }
+    if ft.is_unsigned() && dec.is_negative() {
+        Ok(Decimal::zero())
+    } else {
+        Ok(dec)
+    }
+}
+
+/// `produce_str_with_specified_tp`(`ProduceStrWithSpecifiedTp` in tidb) produces
+/// a new string according to `flen` and `chs`.
+/// # panic
+/// The s must represent a valid str, otherwise, panic!
+fn produce_str_with_specified_tp<'a>(
```

Comment:
Place this function in `codec::convert` module is more appropriate (both `impl_cast.rs` and `builtin_cast.rs` will call it).